File: Others/medium/MinMaxStackConstruction.py
# Feel free to add new properties and methods to the class.
class MinMaxStack:

    def __init__(self):
        self.stack = []
        self.min = None
        self.max = None

    def peek(self):
        if self.stack:
            return self.stack[-1][0]

    def pop(self):
        if self.stack:
            num = self.stack.pop()[0]
            if self.stack:
                _, next_min, next_max = self.stack[-1]

                if next_min != self.min:
                    self.min = next_min
                if next_max != self.max:
                    self.max = next_max
            else:
                self.min = None
                self.max = None

            return num

    def push(self, number):
        if not self.max and not self.min:
            self.min = self.max = number
        else:
            if number > self.max:
                self.max = number
            if number < self.min:
                self.min = number
        self.stack.append((number, self.min, self.max))

    def getMin(self):
        if self.stack:
            return self.min

    def getMax(self):
        if self.stack:
            return self.max

File: Golang/medium/min-max-stack-construction.go
package main

import "math"

type StackElem struct {
	Value, Min, Max int
}

type Stack []StackElem

func (s *Stack) Push(elem StackElem) {
	*s = append(*s, elem)
}

func (s *Stack) Pop() StackElem {
	if len(*s) == 0 {
		panic("Popping from empty stack...")
	}
	elem := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return elem
}

func (s *Stack) Peek() StackElem {
	if len(*s) == 0 {
		panic("Peeking from empty stack...")
	}
	elem := (*s)[len(*s)-1]
	return elem
}

type MinMaxStack struct {
	stack    Stack
	min, max int
}

func NewMinMaxStack() *MinMaxStack {
	return &MinMaxStack{stack: Stack{}, min: math.MaxInt32, max: math.MinInt32}
}

func (stack *MinMaxStack) Peek() int {
	val := stack.stack.Peek()
	return val.Value
}

func (stack *MinMaxStack) Pop() int {
	ret := stack.stack.Pop()
	if len(stack.stack) > 0 {
		val := stack.stack.Peek()
		if val.Max != stack.max {
			stack.max = val.Max
		}
		if val.Min != stack.min {
			stack.min = val.Min
		}
	}
	return ret.Value
}

func (stack *MinMaxStack) Push(number int) {
	if number > stack.max || len(stack.stack) == 0 {
		stack.max = number
	}
	if number < stack.min || len(stack.stack) == 0 {
		stack.min = number
	}
	stack.stack.Push(StackElem{Value: number, Min: stack.min, Max: stack.max})
}

func (stack *MinMaxStack) GetMin() int {
	return stack.min
}

func (stack *MinMaxStack) GetMax() int {
	return stack.max
}
